### PR TITLE
Unknown injections handling in offline

### PR DESCRIPTION
**Pull request by GarethCabournDavies**
Allow the offline search to use injections which do not have an associated injection file.

For the use-case I am thinking of, we would be provided frame files with an unknown set of injections by an external team.

We would then use the alternative datafind settings on the config for the injected data, but use the original uninjected data for the background in the same way we do for known injections.

These would then show up as an additional minifollowup in the open box results, and a `.hdf` results file is made in the open box results directory which can be used for reporting results

Draft status is due to ongoing testing, this will be removed once the testing has completed and passed sensible checks

### CODE REVIEW

**Comment by GarethCabournDavies** [comment]
Draft status removed: I had been wanting to allow in-frame injections which have a file as well, in case of supplied known injections, but I realised that can be done using #4363 and overriding the `--injection-scale-factor` option to be (close to) infinite for those injections, similar to `single_template-inj_params_noinj` plotting option, so no development is needed

**Review by spxiwh — APPROVED** [review]
I think I'm happy with this. Is there an example output page or test?

File: bin/workflows/pycbc_make_coinc_search_workflow
Comment (GarethCabournDavies) [reply bin/workflows/pycbc_make_coinc_search_workflow:472]: Thanks for the pointer, missed that!
Added now, and running test workflow to generate some results pages. I'll add some injection generation / searching into the CI tests as well (similar to the live CI)

**Comment by GarethCabournDavies** [comment]
Note that changes since the approval will show up on  https://github.com/GarethCabournDavies/pycbc/compare/8582d66ad67574965e47d6519faf9e8a45ebbc56...unknown_injections

**Comment by GarethCabournDavies** [comment]
Note: Current failure in the CI will be fixed by #4404

**Comment by spxiwh** [comment]
Poke: Still failing search test.

**Comment by GarethCabournDavies** [comment]
Searched through the logs and can't find _any_ reasons why the coinc jobs for unknown injections are failing, just the transfers of the output file

**Comment by GarethCabournDavies** [comment]
Okay, the CI failures are due to huge injinj / fullinj / injfull jobs which are being booted I think - at least they are taking a lot of time and memory on my local runs using the examples config

I think that merging #4420 will fix the issues, so unfortunately, that becomes another dependency!

**Comment by spxiwh** [comment]
Looks like there's now conflicts here. Poke me when this is ready!

**Comment by GarethCabournDavies** [comment]
Okay @spxiwh, this looks sensible on the results pages now - I thought the wrong data was being passed to the minifollowups, but it was me trying to overwrite rather than use a fresh directory

I'm _slightly_ worried that the config file syntax becomes very complicated at this point, but I suppose this is a non-standard use-case

File: pycbc/io/hdf.py
Comment (spxiwh) [inline pycbc/io/hdf.py:1046]: Is the `self_inclusive` flag from before not still relevant or active? Does it need removing in any way?

File: pycbc/workflow/injection.py
Comment (spxiwh) [inline pycbc/workflow/injection.py:255]: My only substantial comment is that having a File be None is a hack. If anyone accidentally sent this file to a job, pegasus is going to [rightly] break. There doesn't seem to be a good way to handle this though ...

**Review by spxiwh — COMMENTED** [review]
Thanks. I think I'm largely okay with this, do see some of the comments though.

File: bin/plotting/pycbc_page_foreground
Comment (GarethCabournDavies) [reply bin/plotting/pycbc_page_foreground:27]: That's right, this executable is used for the foreground, but needs the exclusive to be added so that it _can_ be used in the injection case (and the title allow us to state that this is not the foreground case)

File: bin/workflows/pycbc_make_offline_search_workflow
Comment (GarethCabournDavies) [reply bin/workflows/pycbc_make_offline_search_workflow:583]: That was a "Considered error" that I forgot to revert - reverted now

File: bin/workflows/pycbc_make_offline_search_workflow
Comment (GarethCabournDavies) [reply bin/workflows/pycbc_make_offline_search_workflow:478]: Error, reverted

File: pycbc/io/hdf.py
Comment (GarethCabournDavies) [reply pycbc/io/hdf.py:1046]: I missed the comment at the top of the function, stating to use has_inc=False for injections, so double-included this

File: examples/search/plotting.ini
Comment (GarethCabournDavies) [reply examples/search/plotting.ini:3]: reverted

File: examples/search/analysis.ini
Comment (GarethCabournDavies) [reply examples/search/analysis.ini]: changed

File: bin/workflows/pycbc_make_offline_search_workflow
Comment (GarethCabournDavies) [reply bin/workflows/pycbc_make_offline_search_workflow]: fixed

File: examples/search/gen.sh
Comment (GarethCabournDavies) [reply examples/search/gen.sh:12]: I can revert the changes to the CI after final approval, if that is best?

Handwaving discussion over time though:
 - the workflow additions are all parallel to already-running jobs
 - the generation of data with injections does take a few minutes of added time before the workflow starts, and is not parallel

File: pycbc/workflow/injection.py
Comment (GarethCabournDavies) [reply pycbc/workflow/injection.py:255]: Yeah, though people _shouldn't_ do that, there wont be a nice error message to say what they have done which is wrong either

File: examples/search/gen.sh
Comment (spxiwh) [reply examples/search/gen.sh:12]: The CI only has a single machine, so cannot parallelize. It's going to have to run all the jobs here one by one (or perhaps two-by-two). Therefore we do need to know how much extra time this is adding, and if this feature is "niche" enough to warrant this, given that there's probably many possibilities in workflow that we are not testing.

**Comment by spxiwh** [comment]
Okay, I'm going to remove this from my "to-do" list now. Please do consider CI time, but merge this once you are happy and the CI tests (including the search workflow!) have completed. If you make any more substantive changes, explicitly poke me again.

**Comment by tdent** [comment]
@spxiwh is the use case for this going to come up in the near future to your knowledge?  If so, we will want this to be on master

**Comment by GarethCabournDavies** [comment]
This will need a substantial re-think, as the use-case I've implemented here is unlikely to be used in the near future

What we are going to need is the case where injection frame files are supplied, and an injection definition file is _also_ given, but we are not injecting them ourselves.

As a result, I will close, and think about how to adapt this for that use-case.

**Pull request closed without merging**
